### ISSUE TITLE
Incorrect type narrowing on assignment when assigned type has invariant type param and target type contains `Any`

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#5628]
**Describe the bug**
```python
from typing import Any, Generic, Iterable, TypeVar, Union, cast

T = TypeVar("T")


class Foo(Generic[T]):
    ...


def bar(s: Iterable[int]):
    ...


def foo(x: Iterable[Union[int, Foo[Any]]]):
    x = cast(list[int], x)
    reveal_type(x)  # list[int | Foo[Any]] strangely. Cast only kind of worked.
    bar(x) # False positive type error here.
```

I'd expect no type error on last line, but cast seems to be ignored here. Foo being generic is important for this example and simplifying it to `class Foo:` fixes the error.

**VS Code extension or command-line**
pyright 1.1.320. This looks to be recent bug. 1.1.318 type checks and 1.1.319 is when this started failing.

**Comment by erictraut** [comment]
This is working as designed. The type of the expression `cast(list[int], x)` is `list[int]` due to the cast. This value is assigned back to the variable `x`, whose type is declared as `Iterable[int | Foo[Any]]`, but the type narrowing is applied upon assignment, so the type is narrowed.

If you want the cast to "stick", you need to assign the value back to a variable whose type is note declared.

```python
def func(x: Iterable[Union[int, Foo[Any]]]):
    x2 = cast(list[int], x)
    reveal_type(x2)  # list[int]
    bar(x2)
```

**Comment by erictraut** [comment]
Actually, the narrowing that's being applied here doesn't look right. Reopening for further investigation.

```python
def func(y: list[int]):
    x: Iterable[int | Foo[Any]] = y
    reveal_type(x) # list[int | Foo[Any]]
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.321, which I just published. It will also be included in a future release of pylance.
